### PR TITLE
Add support for removing assets from Vault

### PR DESCRIPTION
**Pull request by shahthepro**
# Changes

* Added `removeAsset` to `VaultAdmin`

## Gas impact

Both the following are after the Frax strategy has been removed.
The before is when frxETH is still an asset of the Vault.
After has the frxETH asset removed.

| Test |  Before | After | Saving |
| --- | --- | --- | --- |
mint using 1 WETH | 531,368 | 512,991 | 18,377 |
mint using 11 WETH | 745,615 |  717,792 | 27,823 |
redeem 10 OETH | 574,376 |  555,992 | 18,377 |
redeem 100 OETH | 574,400 | 556,016 | 18,377 |
rebase | 467,203 | 448,826 | 18,377 |

# Security

If you made a contract change, make sure to complete the checklist below before merging it in master.

Refer to our [documentation](https://github.com/OriginProtocol/security) for more details about contract security best practices.

Contract change checklist:
  - [x] Code reviewed by 2 reviewers. 
  - [x] Copy & paste code review [security checklist](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md) below this checklist.
  - [x] Unit tests pass
  - [x] Slither tests pass with no warning
  - [x] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

File: contracts/contracts/vault/VaultAdmin.sol [hunk 1]
```diff
@@ -343,6 +343,59 @@ contract VaultAdmin is VaultStorage {
...
+            IVault(address(this)).checkBalance(_asset) == 0,
+            "Vault still holds asset"
+        );
+
+        uint256 assetsCount = allAssets.length;
+        uint256 assetIndex = assetsCount;
+        for (uint256 i = 0; i < assetsCount; ++i) {
+            if (allAssets[i] == _asset) {
+                assetIndex = i;
+                break;
```
Comment (DanielVF) [inline contracts/contracts/vault/VaultAdmin.sol:358]: We should really revert after this loop if the asset is not found. In theory anything with isSupported will be the allAssets, but this could break stuff badly if it were not.

File: contracts/contracts/vault/VaultAdmin.sol [hunk 2]
```diff
@@ -343,6 +343,59 @@ contract VaultAdmin is VaultStorage {
...
+        assets[_asset].isSupported = false;
+
+        emit AssetRemoved(_asset);
+    }
+
+    /**
+     * @notice Remove a supported asset from the Vault
+     * @param _asset Address of asset
+     */
+    function removeAsset(address _asset) external onlyGovernor {
```
Comment (DanielVF) [inline contracts/contracts/vault/VaultAdmin.sol:385]: Let's just have the one `removeAsset`, and not have `removeAssets`. We're not likely to call this often enough to justify the extra code size in the contract.

Comment (shahthepro) [reply contracts/contracts/vault/VaultAdmin.sol:385]: Will do it

File: contracts/contracts/vault/VaultAdmin.sol (hunk 1 above)
Comment (shahthepro) [reply contracts/contracts/vault/VaultAdmin.sol:358]: Actually, `assetIndex` is set to `assetsCount` initially. So it'll revert if the asset doesn't exist `allAssets` with an out-of-bound index error. Have also added a comment about this (a few lines below this)

File: contracts/contracts/vault/VaultAdmin.sol [hunk 3]
```diff
@@ -343,6 +343,45 @@ contract VaultAdmin is VaultStorage {
...
+        // Update allAssets array
+        allAssets[assetIndex] = allAssets[assetsCount - 1];
+        allAssets.pop();
+
+        // Reset default strategy
+        assetDefaultStrategies[_asset] = address(0);
+        emit AssetDefaultStrategyUpdated(_asset, address(0));
+
+        // Mark as unsupported
+        assets[_asset].isSupported = false;
```
Comment (DanielVF) [inline contracts/contracts/vault/VaultAdmin.sol:380]: Let's actually clear everything in the asset struct with `delete`, rather than just isSupported. Feels cleaner and will keep us from having weird data around lurking for the future if we ever decided to turn the asset back on.

Comment (shahthepro) [reply contracts/contracts/vault/VaultAdmin.sol:380]: Sounds good. Will make that change

File: contracts/contracts/vault/VaultAdmin.sol (hunk 1 above)
Comment (DanielVF) [reply contracts/contracts/vault/VaultAdmin.sol:362]: Clever. I've pushed up one additional code comment on this.

File: contracts/contracts/vault/VaultAdmin.sol [hunk 4]
```diff
@@ -343,6 +343,45 @@ contract VaultAdmin is VaultStorage {
...
     }
 
+    /**
+     * @notice Remove a supported asset from the Vault
+     * @param _asset Address of asset
+     */
+    function removeAsset(address _asset) external onlyGovernor {
+        require(assets[_asset].isSupported, "Asset not supported");
+        require(
+            IVault(address(this)).checkBalance(_asset) == 0,
```
Comment (DanielVF) [inline contracts/contracts/vault/VaultAdmin.sol:353]: I'm wondering if we are actually going to be able to hit zero on some assets that are weird like stETH. Maybe this should be < 1e6 to allow for dust.

**Review by DanielVF — COMMENTED** [review]
# Code review remove assets

## Requirements

We will be removing some assets from OETH. We want the data structures in the contract to be like these assets were never there.

## Easy Checks

#### Authentication

- [x] Never use tx.origin
- [x] Every external/public function is supposed to be externally accessible
- [x] Every external/public function has the correct authentication

#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [x] Contract has no payable methods.

#### Cryptographic code

_no crypto_

#### Gas problems

- [x] Contracts with for loops must have either:
    - [x] A way to remove items
    - [x] Can be upgraded to get unstuck
    - [x] Size can only controlled by admins
- [x] Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [x] Code is solidity version >= 0.8.0
- [x] All for loops use uint256

#### Proxy

- [x] No storage variable initialized at definition when contract used as a proxy implementation.

#### Events
- [x] All state changing functions emit events

## Medium Checks

#### Rounding
_No rounding_

#### Dependencies
_No dependency changes_

#### External calls

- [x] Contract addresses passed in are validated
- [x] No unsafe external calls
- [x] No Malicious behaviors
- [x] Low level call() must require success.
- [x] No slippage attacks (we need to validate expected tokens received)
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Do not call balanceOf for external contracts to determine what they will do when they use internal accounting

#### Tests

  - [x] Each publicly callable method has a test
  - [ ] Each logical branch has a test
  - [ ] Each require() has a test
  - [ ] Edge conditions are tested

#### Deploy

- [x] Deployer permissions are removed after deploy

#### Downstream

_No downstream_

## Thinking

#### Logic

The code needs to remove the asset from all storage that holds the asset config.

- assets ✅
- allAssets ✅
- assetDefaultStrategies ✅

We also have sanity code to ensure that the vault already has been emptied of the asset before removing it.

#### Deployment Considerations

_No deployment considerations_

#### Internal State

See above under "Logic"

#### Attack

Removing the wrong asset could result in an apparent drop in the vaults assets, thus scaring users and stopping all redeems and rebases. Also would stop mints in that asset. 

However, we do check that the asset is actually unused before allowing the removal of an asset.

We also check that the asset is a supported asset, and is in the assets array before removing.

This action is governance only, and so should not be available for external users to attack.

This method does not make external, non-static calls. It should be safe from any kind of reentrancy attack happening inside it.

It is not marked non-reentrant, so it could be triggered externally. Even though it is a governance action that can only be called from the timelock, the governance actions that have come out of queue in the timelock can be triggered by anyone, thus allowing an attacker to trigger a governance queued remove.

Removing an asset results in that asset not being available for withdraw and a lower total vault value. Neither of these are good for an attacker. Removing an asset during an OUSD redeem would cause a revert since the asset for loop would cause a revert. For both mints and redeems, we do all calculations first, before calling external assets.

🟣 An attacker could break the strategy remove by front running a donation to the vault. I think we'll just deal with this if it happens with an upgrade.



#### Flavor

Flavor great. Code is simple!

File: contracts/contracts/vault/VaultAdmin.sol [hunk 5]
```diff
@@ -343,6 +343,45 @@ contract VaultAdmin is VaultStorage {
...
     }
 
+    /**
+     * @notice Remove a supported asset from the Vault
+     * @param _asset Address of asset
+     */
+    function removeAsset(address _asset) external onlyGovernor {
+        require(assets[_asset].isSupported, "Asset not supported");
+        require(
+            IVault(address(this)).checkBalance(_asset) <= 1e6,
```
Comment (pandadefi) [inline contracts/contracts/vault/VaultAdmin.sol:353]: I saw the original version had a check for zero balance, considering you might have dust on multiple strategies as well 1e6 might be a little low, and would advise for 10e12 which is even for a coin like ETH not worth more than a cent.

**Review by pandadefi — APPROVED** [review]

File: contracts/contracts/vault/VaultAdmin.sol (hunk 5 above)
Comment (shahthepro) [reply contracts/contracts/vault/VaultAdmin.sol:353]: Sounds good to me, making that change

Comment (shahthepro) [reply contracts/contracts/vault/VaultAdmin.sol]: Done

File: contracts/contracts/vault/VaultAdmin.sol (hunk 4 above)
Comment (shahthepro) [reply contracts/contracts/vault/VaultAdmin.sol]: Panda suggested to increase it to 1e13 (since the dust can exist across multiple strategies and it'd still be insignificant), so used that limit

File: contracts/contracts/vault/VaultAdmin.sol (hunk 3 above)
Comment (shahthepro) [reply contracts/contracts/vault/VaultAdmin.sol]: Done

**Review by DanielVF — DISMISSED** [review]

File: contracts/contracts/vault/VaultAdmin.sol [hunk 6]
```diff
@@ -343,6 +343,45 @@ contract VaultAdmin is VaultStorage {
...
     }
 
+    /**
+     * @notice Remove a supported asset from the Vault
+     * @param _asset Address of asset
+     */
+    function removeAsset(address _asset) external onlyGovernor {
+        require(assets[_asset].isSupported, "Asset not supported");
+        require(
+            IVault(address(this)).checkBalance(_asset) <= 1e13,
```
Comment (naddison36) [inline contracts/contracts/vault/VaultAdmin.sol:353]: why an external call and not an internal `_checkBalance(_asset)`?

File: contracts/deploy/092_vault_admin_upgrade.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    Part of simplified OETH proposal. This adds the ability to remove a supported asset from the Vault. \
+    ",
+      actions: [
+        {
+          contract: cVaultAdmin,
+          signature: "setAdminImpl(address)",
+          args: [dVaultAdminImpl.address],
+        },
+      ],
+    };
```
Comment (naddison36) [inline contracts/deploy/092_vault_admin_upgrade.js:38]: can we add the removal of frxETH to this deployment?

**Review by naddison36 — COMMENTED** [review]
It'd be interesting to see the gas impacts of mint, redeem and rebase when frxETH is removed.

File: contracts/contracts/vault/VaultAdmin.sol (hunk 6 above)
Comment (shahthepro) [reply contracts/contracts/vault/VaultAdmin.sol:353]: `checkBalance` is in `VaultCore`, not `VaultAdmin`

File: contracts/deploy/092_vault_admin_upgrade.js (hunk 7 above)
Comment (shahthepro) [reply contracts/deploy/092_vault_admin_upgrade.js:38]: We can do it, but we need strategist to do swap the frxETH it holds. Right now, this PR allows dust upto `e12`. But the Vault right now has a little over `1e16` frxETH

File: contracts/test/vault/oeth-vault.fork-test.js [hunk 8]
```diff
@@ -148,7 +148,36 @@ describe("ForkTest: OETH Vault", function () {
       await expect(tx).to.revertedWith("Liquidity error");
     });
     it("OETH whale can redeem after withdraw from all strategies", async () => {
-      const { oeth, oethVault, timelock } = fixture;
+      const { oeth, oethVault, timelock, domen, strategist, frxETH } = fixture;
```
[the remark this answers is not in the document]
Comment (shahthepro) [reply contracts/test/vault/oeth-vault.fork-test.js:151]: frxETH strategy is already removed. It's just the asset that we have to remove right now

File: contracts/contracts/vault/VaultAdmin.sol (hunk 6 above)
Comment (naddison36) [reply contracts/contracts/vault/VaultAdmin.sol:353]: Ah, yes, that makes sense.

**Review by naddison36 — DISMISSED** [review]

File: contracts/deploy/092_vault_admin_upgrade.js (hunk 7 above)
Comment (naddison36) [reply contracts/deploy/092_vault_admin_upgrade.js]: I've added the remove of frxETH to the deploy script as the remaining frxETH in the Vault has been swapped out for WETH
